### PR TITLE
Send activity tick by polling

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Addresses
https://issues.redhat.com/browse/ODC-6644

Sends tick by polling even when drawer is minimised

# Screenshots
Screenshot 2022-04-25 at 8 14 09 PM

Screenshot 2022-04-25 at 8 15 09 PM
4.09 PM.png…]()

# Tests
no change

# Browser conformance

Chrome

### CODE REVIEW

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 1]
```diff
@@ -74,13 +76,26 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
...
+      activityTick();
+      tickOnInterval = tickOnInterval || setInterval(activityTick, 30000);
+    } else if (tickOnInterval) {
+      clearInterval(tickOnInterval);
+      tickOnInterval = null;
+    }
+  }
+
+  document.addEventListener('visibilitychange', handleTickOnVisible, false);
+  handleTickOnVisible();
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:94]: This should be done in a useEffect and remove the event listener on unmount

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 2]
```diff
@@ -74,13 +76,26 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
...
   const tick = useActivityTick(workspaceName, namespace);
 
-  const onData = React.useCallback(
-    (data: string): void => {
-      tick();
-      ws.current?.send(`0${Base64.encode(data)}`);
-    },
-    [tick],
-  );
+  const activityTick = React.useCallback(() => tick(), [tick]);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:79]: tick is already wrapped in a `useCallback`

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 3]
```diff
@@ -74,13 +76,34 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
...
 
-  const onData = React.useCallback(
-    (data: string): void => {
-      tick();
-      ws.current?.send(`0${Base64.encode(data)}`);
-    },
-    [tick],
-  );
+  React.useEffect(() => {
+    let tickOnInterval = setInterval(tick, 30000);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:80]: interval should be `60000`

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 4]
```diff
@@ -74,13 +76,34 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
...
+
+    // Function to handle ticking
+    const handleTickOnVisible = () => {
+      if (document.visibilityState === 'hidden' || !isActiveTab) {
+        clearInterval(tickOnInterval);
+        tickOnInterval = null;
+      }
+      if (document.visibilityState === 'visible' && isActiveTab) {
+        tickOnInterval = tickOnInterval || setInterval(tick, 60000);
+        tick(); // Start with  a sync tick
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:90]: Why do you need to call tick() seperately

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx (hunk 3 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx]: updated

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx (hunk 4 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx]: updated

**Comment by karthikjeeyar** [comment]
can we have some unit tests for this `CloudShellExec` component?

**Comment by abhinandan13jan** [comment]
@karthikjeeyar added test for CloudShellExec

**Comment by karthikjeeyar** [comment]
Verified on clusterbot
image

/lgtm
/label qe-approved

**Comment by sahil143** [comment]
/lgtm

verified on cluster bot

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 5]
```diff
@@ -74,13 +76,33 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
...
+
+    document.addEventListener('visibilitychange', handleTickOnVisible, false);
+    handleTickOnVisible();
+
+    return () => {
+      clearInterval(tickOnInterval);
+      tickOnInterval = null;
+      document.removeEventListener('visibilitychange', handleTickOnVisible);
+    };
+  }, [isActiveTab, tick]);
```
Comment (invincibleJai) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:101]: will it make sense to use `requestAnimationFrame` here? can avoid listeners on `visibilitychange` and `setInterval`.

WDYT? 

cc @sahil143 @karthikjeeyar 

```suggestion
  React.useEffect(() => {
   let startTime;
   const handleTick = (timestamp) => {
   if ((!startTime || timestamp - startTime >= TICK_INTERVAL) && isActiveTab) {
     startTime = timestamp;
     tick();
    }
    requestAnimationFrame(handleTick);
   }
  
  requestAnimationFrame(handleTick);
  
  return () => {
    cancelAnimationFrame(handleTick);
  }
  }, [isActiveTab, tick]);
```

Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:101]: Hi @invincibleJai , not able to unit test this out correctly.

Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:101]: setTimeout is unit testable at jest. Although we are using a debounce but the requestAnimationFrame is triggered at around 60fps, do you think it is a good fit for the case here?

**Comment by abhinandan13jan** [comment]
cc: @invincibleJai  Thanks for the solution. tested on clusterbot. Works fine
Screenshot 2022-05-17 at 10 41 19 PM

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx [hunk 6]
```diff
@@ -74,13 +76,27 @@ const CloudShellExec: React.FC<CloudShellExecProps> = ({
...
+      }
+      tickReq = window.requestAnimationFrame(handleTick);
+    };
+
+    tickReq = window.requestAnimationFrame(handleTick);
+
+    return () => {
+      window.cancelAnimationFrame(tickReq);
+    };
+  }, [isActiveTab, tick]);
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:95]: Every tab will be running this `requestAnimationFrame` loop at all times even if it is not the active tab. You could alleviate this by not requesting a frame unless it is the active tab. 

That being said, the `tick` isn't specific to a tab and could be moved higher in the component tree providing we know higher up when we have a connection.

Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:95]: @christianvogt Let us know how to proceed?
1. We could have this call higher up at the `MultiTabTerminal` level.  But the problem is the `activityTick('workspacename', 'namespace')`. This `workspaceName`, `namespace' data is only available at an individual terminal Level. To have this data higher up would require a lot of refactoring.

2. assign tickReq only for the activeTab else assign null. something like `tickReq = isActive ?  window.requestAnimationFrame(handleTick) : null ` 
WDYT?
cc: @sahil143 @invincibleJai

Comment (christianvogt) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellExec.tsx:95]: We only want to send the tick if there's an open connection. It's either each terminal is responsible for sending the tick when active. Or each terminal is responsible for issuing a callback of its status to the parent which can then control sending the tick with the information provided by its terminal children. Having a single "manager" that is responsible for the tick would be preferred but both options work.

I didn't look if there's any performance implications of the current implementation. I assumed that each terminal would be running a tight animation loop, therefore doing _some_ work at all times; but it's not a heavy load.

**Pull request merged**